app/controllers/katello/api/registry/registry_proxies_controller.rb

```diff
@@ -210,94 +205,67 @@ def pull_blob
-
-      manifest = create_manifest
-      return if manifest.nil?
-
-      begin
-        files = get_manifest_files(repository, manifest)
-        return if files.nil?
-
-        tar_file = create_tar_file(files, repository, tag)
-        return if tar_file.nil?
-
-        digest = upload_manifest(tar_file)
-        return if digest.nil?
-
-        tag = upload_tag(digest, tag)
-        return if tag.nil?
-      ensure
-        File.delete(tmp_file('manifest.json')) if File.exist? tmp_file('manifest.json')
+      headers = translated_headers_for_proxy
+      headers['Content-Type'] = request.headers['Content-Type'] if request.headers['Content-Type']
+      body = @_request.body.read
+      pulp_response = Resources::Registry::Proxy.put(@_request.fullpath, body, headers)
+      pulp_response.headers.each do |key, value|
+        response.header[key.to_s] = value
       end
-
-      render json: {}
-    end
-
-    # FIXME: This is referring to a non-existent Pulp 2 server.
-    # Pulp 3 container push support is needed instead.
-    def pulp_content
-      Katello.pulp_server.resources.content
+      head pulp_response.code
     end
 
     def start_upload_blob
-      uuid = SecureRandom.hex(16)
-      response.header['Location'] = "#{request_url}/v2/#{params[:repository]}/blobs/uploads/#{uuid}"
-      response.header['Docker-Upload-UUID'] = uuid
-      response.header['Range'] = '0-0'
-      head 202
-    end
+      headers = translated_headers_for_proxy
+      headers['Content-Type'] = request.headers['Content-Type'] if request.headers['Content-Type']
+      headers['Content-Length'] = request.headers['Content-Length'] if request.headers['Content-Length']
+      pulp_response = Resources::Registry::Proxy.post(@_request.fullpath, @_request.body, headers)
 
-    def status_upload_blob
-      response.header['Location'] = "#{request_url}/v2/#{params[:repository]}/blobs/uploads/#{params[:uuid]}"
-      response.header['Range'] = "123"
-      response.header['Docker-Upload-UUID'] = "123"
-      render plain: '', status: :no_content
+      pulp_response.headers.each do |key, value|
+        response.header[key.to_s] = value
+      end
+      head pulp_response.code
     end
 
-    def chunk_upload_blob
-      response.header['Location'] = "#{request_url}/v2/#{params[:repository]}/blobs/uploads/#{params[:uuid]}"
-      render plain: '', status: :accepted
+    def translated_headers_for_proxy
+      current_headers = {}
+      env = request.env.select do |key, _value|
+        key.match("^HTTP_.*")
+      end
+      env.each do |header|
+        current_headers[header[0].split('_')[1..-1].join('-')] = header[1]
+      end
+      current_headers
     end
 
     def upload_blob
-      File.open(tmp_file("#{params[:uuid]}.tar"), 'ab', 0600) do |file|
-        file.write request.body.read
+      headers = translated_headers_for_proxy
+      headers['Content-Type'] = request.headers['Content-Type'] if request.headers['Content-Type']
```

Comment:
It really feels like there should be a more concise way to do all these calls but neither ||= or &&= works. Perhaps you could split all these test blocks into a method?

Reply:
Lemme take a look about if there's a less verbose way to do this that doesn't lose readability.

Reply:
Based on our video call I'm happy to call this ok. It's very readable and quick as you wrote it.

Reply:
I didn't find a less verbose way to write this, that check for nil seems like a requirement. C'mon Ruby!